human-readable-types/src/main/java/com/palantir/humanreadabletypes/StringBasedUrl.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * (c) Copyright 2020 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.humanreadabletypes;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonValue;
+import java.net.MalformedURLException;
+import java.net.URL;
+
+/**
+ * An immutable {@link URL} whose {@link #equals}, {@link #toString}, and {@link #hashCode} are derived from the URL's
+ * string representation (at instance construction time) rather than the {@link URL} class. For example, two {@link
+ * StringBasedUrl}s are equal iff the strings from which they were {@link #valueOf constructed} are equal.
+ * <p>
+ * This class is intended to be used in configuration files where (1) host name resolution is not desired, and (2)
+ * equality/stability of the configuration file should be derived from the configuration file syntax, rather than the
+ * dynamic behavior of the configuration values, cf. {@link URL#equals}.
```

Comment:
Can you clarify why one would want to use this class over `java.net.URI`?

Reply:
because all sorts of things are valid URIs, including `""`. I suspect that 99.99% of things (at least!) that people put in config are URLs, hence we can make the check more specific.